nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/TestSQLRepositoryAPI.java

```diff
@@ -4431,26 +4438,173 @@ public void testLockingBeforeSave() {
+        acl.add(new ACE("bob", "Write", true));
+        acp.addACL(acl);
+        folder.setACP(acp, true);
+        DocumentModel subFolder = session.createDocumentModel("/fold", "subfold", "Folder");
+        session.createDocument(subFolder);
+        DocumentModel doc = session.createDocumentModel("/fold/subfold", "doc", "File");
+        doc = session.createDocument(doc);
+
+        // must be a record
+        try {
+            session.setRetentionUntil(doc.getRef(), null);
+            fail("Should fail because document is not a record");
+        } catch (PropertyException e) {
+            assertEquals("Document is not a record", e.getMessage());
+        }
+
+        // make it a record
+        session.makeRecord(doc.getRef());
+        // bob can modify it
+        try (CloseableCoreSession bobSession = openSessionAs("bob")) {
+            assertTrue(bobSession.hasPermission(doc.getRef(), SecurityConstants.WRITE_PROPERTIES));
+        }
+
+        // set legal hold
+        assertFalse(session.hasLegalHold(doc.getRef()));
+        session.setLegalHold(doc.getRef(), true);
+        assertTrue(session.hasLegalHold(doc.getRef()));
+
+        checkUndeletable(folder, doc);
+        // bob cannot modify it
+        try (CloseableCoreSession bobSession = openSessionAs("bob")) {
+            assertFalse(bobSession.hasPermission(doc.getRef(), SecurityConstants.WRITE_PROPERTIES));
+        }
+
+        // unset legal hold, can now delete
+        session.setLegalHold(doc.getRef(), false);
+        // bob can modify it
+        try (CloseableCoreSession bobSession = openSessionAs("bob")) {
+            assertTrue(bobSession.hasPermission(doc.getRef(), SecurityConstants.WRITE_PROPERTIES));
+        }
+        session.removeDocument(doc.getRef());
+    }
+
+    @Test
+    public void testRetentionExpiresAutomatically() throws Exception {
+        DocumentModel folder = session.createDocumentModel("/", "fold", "Folder");
+        folder = session.createDocument(folder);
+        DocumentModel subFolder = session.createDocumentModel("/fold", "subfold", "Folder");
+        session.createDocument(subFolder);
+        DocumentModel doc = session.createDocumentModel("/fold/subfold", "doc", "File");
+        doc = session.createDocument(doc);
+        session.makeRecord(doc.getRef());
+
+        // set retention to two seconds in the future
+        Calendar twoSeconds = Calendar.getInstance();
+        twoSeconds.add(Calendar.SECOND, 2);
+        session.setRetentionUntil(doc.getRef(), twoSeconds);
+        session.save();
+
+        checkUndeletable(folder, doc);
+
+        // wait 3s to pass retention expiration date
+        nextTransaction();
+        Thread.sleep(3_000);
+        // trigger manually instead of waiting for scheduler
+        new RetentionExpiredFinderListener().handleEvent(null);
+        // wait for all bulk commands to be executed
+        assertTrue("Bulk action didn't finish", bulkService.await(Duration.ofSeconds(60)));
```

Comment:
`nextTransaction`below will wait for bulk processing for at most 3 minutes, you can get rit of this line, bulkService import etc...

Reply:
No this is not the one from `TransactionalFeature`...